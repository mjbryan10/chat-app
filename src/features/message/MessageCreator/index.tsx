import React from 'react'
import InputSubmitButton from 'components/InputSubmitButton'

const MessageCreator = () => {
  return (
    <form>
      <textarea />
      <InputSubmitButton value="Send" />
    </form>
  )
}

export default MessageCreator
