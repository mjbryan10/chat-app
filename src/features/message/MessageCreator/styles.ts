import styled, { ThemeProps, Theme } from 'styled-components';

export const TextArea = styled.textarea<ThemeProps<Theme>>`
   min-height: 75px;
   font-family: inherit;
   font-size: 1rem;
   padding: 0.3rem;
   /* min-width: 80%; */
   flex-grow: 1;
   margin: 15px 15px 0 0;
   border-radius: 0;
   border: ${({ theme }) => theme.border};
   background: ${({ theme }) => theme.background.inputField};
   color: ${({ theme }) => theme.color.text ?? 'inherit'};
   &::placeholder {
      color: ${({ theme }) => theme.color.text ?? 'inherit'};
      opacity: 0.6;
   }
`;

export const Form = styled.form`
  display: flex;
  width: 100%;
  align-items: center;
`;