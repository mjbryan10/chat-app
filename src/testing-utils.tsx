import React from 'react';
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { dark } from './shared/theme';

const AllTheProviders = ({ children }: any) => {
  return (
    <ThemeProvider theme={dark}>
        {children}
    </ThemeProvider>
  )
  
}

const customRender = (ui?: any, options?: any) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }