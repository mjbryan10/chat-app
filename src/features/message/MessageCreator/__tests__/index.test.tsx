import React from 'react';
import { render, RenderResult } from 'testing-utils';
import MessageCreator from '..';
import { fireEvent, act, cleanup } from '@testing-library/react';

describe('MessageCreator component', () => {
   const testText = 'This is a test message.';
   let utils: RenderResult;
   beforeEach(() => {
      utils = render(<MessageCreator />);
   });

   test('should be disabled, if disabled === true ', () => {
      cleanup();
      const __handleSubmit = jest.fn();
      utils = render(<MessageCreator disabled={true} handleSubmit={__handleSubmit} />);
      const { getByTestId, getByText } = utils;
      const textbox = getByTestId('message-creator-textbox');
      const sendButton = getByText(/send/i).closest('input');

      act(() => {
         fireEvent.change(textbox, { target: { value: testText } });
         if (sendButton) fireEvent.click(sendButton);
      });

      expect(__handleSubmit).toBeCalledTimes(0);
      expect(textbox).toHaveDisplayValue('');
      expect(textbox).toBeDisabled();
      expect(sendButton).toBeDisabled();
   });

   test('should display a button with the text `Send`', () => {
      const { getByText } = utils;

      expect(getByText(/Send/i)).toBeVisible();
   });

   test('should change value of textbox when typed', () => {
      const { getByTestId } = utils;
      const textbox = getByTestId('message-creator-textbox');

      fireEvent.change(textbox, { target: { value: testText } });

      expect(textbox).toBeVisible();
      expect(textbox).toHaveDisplayValue(testText);
   });

   test('should submit to handler on button click', () => {
      cleanup();
      const __handleSubmit = jest.fn();
      utils = render(<MessageCreator handleSubmit={__handleSubmit} />);
      const { getByTestId, getByText } = utils;
      const textbox = getByTestId('message-creator-textbox');

      const sendButton = getByText(/send/i);

      act(() => {
         fireEvent.change(textbox, { target: { value: testText } });
         fireEvent.click(sendButton);
      });

      expect(__handleSubmit).toBeCalledTimes(1);
      expect(textbox).toHaveDisplayValue('');
   });

   test('should submit to handler on `Enter` keypress', () => {
      cleanup();
      const __handleSubmit = jest.fn();
      utils = render(<MessageCreator handleSubmit={__handleSubmit} />);
      const { getByTestId, getByText } = utils;
      const textbox = getByTestId('message-creator-textbox');

      act(() => {
         fireEvent.change(textbox, { target: { value: testText } });
         fireEvent.keyPress(getByText(testText), {
            key: 'Enter',
            charCode: 13,
            keyCode: 13,
            code: 'Enter',
            shiftKey: null,
         });
      });

      expect(__handleSubmit).toBeCalledTimes(1);
      expect(textbox).toHaveDisplayValue('');
   });
   test('should not submit to handler on `Shift+Enter` keypress', () => {
      cleanup();
      const __handleSubmit = jest.fn();
      utils = render(<MessageCreator handleSubmit={__handleSubmit} />);
      const { getByTestId, getByText } = utils;
      const textbox = getByTestId('message-creator-textbox');

      act(() => {
         fireEvent.change(textbox, { target: { value: testText } });
         fireEvent.keyPress(getByText(testText), {
            key: 'Enter',
            charCode: 13,
            keyCode: 13,
            code: 'Enter',
            shiftKey: true,
         });
      });

      expect(__handleSubmit).toBeCalledTimes(0);
      expect(textbox).toHaveDisplayValue('');
   });

   test('should render correctly', () => {
      const { asFragment } = render(<MessageCreator />);
   
      expect(asFragment()).toMatchSnapshot();
   });
});
