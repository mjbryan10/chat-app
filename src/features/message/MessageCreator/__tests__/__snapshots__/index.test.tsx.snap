// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageCreator component should render correctly 1`] = `
<DocumentFragment>
  <form
    class="sc-AxiKw jlHJyN"
    data-testid="message-creator-form"
  >
    <textarea
      class="sc-AxirZ gaLhWH"
      data-testid="message-creator-textbox"
      placeholder="Type a message"
    />
    <input
      class="sc-AxjAm ernPVN"
      data-testid="input-submit-button"
      type="submit"
      value="Send"
    />
  </form>
</DocumentFragment>
`;
